File: M3 Intermediate DSA/src/L5_Arrays_Prefix_Sum/Assignment3.java
package L5_Arrays_Prefix_Sum;

public class Assignment3 {

  public long[] rangeSum(int[] A, int[][] B) {
    int N = A.length;
    int M = B.length;
    long[] prefixSum = new long[N];
    long[] ans = new long[M];
    prefixSum[0] = A[0];

    for (int index = 1; index < N;index++) {
      prefixSum[index] = prefixSum[index-1] + A[index];
    }

    for (int index = 0; index < M; index++) {
      int left = B[index][0];
      int right = B[index][1];

      long sum = prefixSum[right];

      if (left > 0) {
        sum -= prefixSum[left-1];
      }

      ans[index] = sum;
    }

    return ans;
  }
}

File: M3_Intermediate_DSA/L3_Prefix_Sum/assignment3.go
package main

func getPrefixSum(A []int) []int64 {
	N := len(A)

	prefixSum := make([]int64, N)
	prefixSum[0] = int64(A[0])

	for index := 1; index < N; index++ {
		prefixSum[index] = prefixSum[index-1] + int64(A[index])
	}

	return prefixSum
}

func RangeSum(A []int, B [][]int) []int64 {
	M := len(B)

	prefixSum := getPrefixSum(A)
	ans := make([]int64, M)

	for index := range B {
		leftIndex := B[index][0]
		rightIndex := B[index][1]
		sum := prefixSum[rightIndex]

		if leftIndex > 0 {
			sum -= prefixSum[leftIndex-1]
		}

		ans[index] = sum
	}

	return ans
}
